frontend/src/lib/ForwardMessageCache.ts

```diff
@@ -111,22 +154,41 @@ export class ForwardMsgCache {
       return
     }
 
-    if (!msg.metadata.cacheable || this.messages.has(msg.hash)) {
-      // Don't cache messages that the server hasn't marked as
-      // cacheable, or that we've already cached.
+    if (!msg.metadata.cacheable) {
+      // Don't cache messages that the server hasn't marked as cacheable.
+      return
+    }
+
+    if (this.getCachedMessage(msg.hash, true) !== undefined) {
+      // We've already cached this message; don't need to do
+      // anything more. (Using getCachedMessage() here ensures
+      // that the message's reportRunCount value gets updated as
+      // expected.)
       return
     }
 
     logMessage(`Caching ForwardMsg [hash=${msg.hash}]`)
-    this.messages.set(msg.hash, ForwardMsg.create(msg))
+    this.messages.set(
+      msg.hash,
+      new CacheEntry(ForwardMsg.create(msg), this.reportRunCount))
   }
 
   /**
    * Return a new copy of the ForwardMsg with the given hash
    * from the cache, or undefined if no such message exists.
+   *
+   * If the message's entry exists, its reportRunCount will be
+   * updated to the current value.
    */
-  private getCachedMessage(hash: string): ForwardMsg | undefined {
+  private getCachedMessage(hash: string, updateReportRunCount: boolean): ForwardMsg | undefined {
```

Comment:
Looks like `updateReportRunCount` is always `true`, so this arg isn't needed

Reply:
It's sneakily set to `false` in the unit tests (`ForwardMsgCache.test.ts` line 34)